### DISCUSSION TITLE
[eslint] types are incorrect

### DISCUSSION THREAD

**Discussion by ljharb** [DefinitelyTyped/DefinitelyTyped#56942] [Issues with a @types package]
- v7.2.10 of the eslint types exposes `SourceCode.CursorWithCountOptions`, v7.2.11+ does not. This is a bug and a breaking change in a patch version.
 - SourceCode.RuleContext is missing `getFirstTokens`, which can take a number as the second argument (the `SourceCode.getFirstTokens` type does not allow a number. Not sure if it should or not)

This has blocked eslint-plugin-react from updating the eslint types beyond v7.2.10 for awhile, and I've only just started looking into it.

[a comment by an automated account is not shown]

**Comment by bradzacher** [comment]
> This is a bug and a breaking change in a patch version.

That assumes DT follows semantic versioning - which it does not.
The major/minor are aligned with the package version the types are for, and the patch is incremented with each release of the types.
https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library

EG a breaking types change doesn't mean that v8.x of the types would be releases - it would mean v7.2.11 is released.

Looks like the offending PR #52730 this was acceptable.
I would say that the migration for [your code](https://github.com/yannickcr/eslint-plugin-react/blob/495a4cf85e6e810802e9de1b5f6efe730cb485c7/lib/types.d.ts#L17) would be

```diff
- getFirstTokens(node: estree.Node | ASTNode, options?: eslint.SourceCode.CursorWithCountOptions): eslint.AST.Token[];
+ getFirstTokens: SourceCode.UnaryNodeCursorWithSkipOptions;
```

**Reply by ljharb** [reply]
When i try that, i get type errors because that type doesn’t permit a number as the second argument.

**Reply by bradzacher** [reply]
Looks like someone just needs to PR a `number` overload for this specific function:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint/index.d.ts#L262-L267

I think this would do it?
```diff
        <T extends AST.Token>(
            node: ESTree.Node,
            options:
+               | number
                | ((token: AST.Token) => token is T)
                | { filter: (token: AST.Token) => token is T; includeComments?: false | undefined; count?: number | undefined },
        ): T[];
```

Worth noting that just passing a number is the same as providing an options object `{count: number}`.

**Reply by ljharb** [reply]
Thanks, I’ll give that a shot and make sure tests pass in eslint-plugin-react, and I’ll make the PR once I figure it out.

**Comment by bradzacher** [comment]
> SourceCode.RuleContext is missing getFirstTokens 

It looks like this method is supported by ESLint, but consuming it directly from `context` is deprecated and has been since [v1.3.0 of ESLint - 6 years ago](https://github.com/eslint/eslint/pull/3517). I'm really surprised they haven't bothered to actually break this yet!

Types generally don't expose deprecated features as they aren't used and should be avoided (i.e. adding new usages is bad practice). 

The alternative to this is to consume it from its intended location - `context.getSourceCode().getFirstTokens`

**Reply by ljharb** [reply]
That type unfortunately also doesn’t support a number, despite the deprecated api doing so.

**Reply by bradzacher** [reply]
That would be because of the above issue - the type is incorrect.
